cmd/snap-preseed/preseed_linux.go

```diff
@@ -0,0 +1,188 @@
+	"github.com/snapcore/snapd/timings"
+)
+
+var (
+	// snapdMountPath is where target core/snapd is going to be mounted in the target chroot
+	snapdMountPath = "/tmp/snapd-preseed"
+	syscallMount   = syscall.Mount
+	syscallChroot  = syscall.Chroot
+)
+
+// checkChroot does a basic sanity check of the target chroot environment, e.g. makes
+// sure critical virtual filesystems (such as proc) are mounted. This is not meant to
+// be exhaustive check, but one that prevents running the tool against a wrong directory
+// by an accident, which would lead to hard to understand errors from snapd in preseed
+// mode.
+func checkChroot(preseedChroot string) error {
+	exists, isDir, err := osutil.DirExists(preseedChroot)
+	if err != nil {
+		return fmt.Errorf("cannot verify target chroot directory %s: %v", preseedChroot, err)
+	}
+	if !exists || !isDir {
+		return fmt.Errorf("target chroot directory %s doesn't exist or is not a directory", preseedChroot)
+	}
+
+	// sanity checks of the critical mountpoints inside chroot directory
+	for _, p := range []string{"/sys/kernel/security/apparmor", "/proc/self", "/dev/mem"} {
+		path := filepath.Join(preseedChroot, p)
+		if exists := osutil.FileExists(path); !exists {
+			return fmt.Errorf("target chroot directory validation error: %s doesn't exist", path)
+		}
+	}
+
+	return nil
+}
+
+var seedOpen = seed.Open
+
+var systemSnapFromSeed = func(rootDir string) (string, error) {
+	seedDir := filepath.Join(dirs.SnapSeedDirUnder(rootDir))
+	seed, err := seedOpen(seedDir, "")
+	if err != nil {
+		return "", err
+	}
+
+	// load assertions into temporary database
+	if err := seed.LoadAssertions(nil, nil); err != nil {
+		return "", err
+	}
+
+	tm := timings.New(nil)
+	if err := seed.LoadMeta(tm); err != nil {
+		return "", err
+	}
+
+	// TODO: handle core18, snapd snap.
+	if seed.UsesSnapdSnap() {
+		return "", fmt.Errorf("preseeding with snapd snap is not supported yet")
+	}
+
+	var coreSnapPath string
+	ess := seed.EssentialSnaps()
+	if len(ess) > 0 {
+		if ess[0].SnapName() == "core" {
+			coreSnapPath = ess[0].Path
+		}
+	}
+
+	if coreSnapPath == "" {
+		return "", fmt.Errorf("core snap not found")
+	}
+
+	return coreSnapPath, nil
+}
+
+func prepareChroot(preseedChroot string) (func(), error) {
+	if err := syscallChroot(preseedChroot); err != nil {
+		return nil, fmt.Errorf("cannot chroot into %s: %v", preseedChroot, err)
+	}
+
+	if err := os.Chdir("/"); err != nil {
+		return nil, fmt.Errorf("cannot chdir to /: %v", err)
+	}
+
+	// GlobalRootDir is now relative to chroot env. We assume all paths
+	// inside the chroot to be identical with the host.
+	rootDir := dirs.GlobalRootDir
+	if rootDir == "" {
+		rootDir = "/"
+	}
+
+	coreSnapPath, err := systemSnapFromSeed(rootDir)
+	if err != nil {
+		return nil, err
+	}
+
+	// create mountpoint for core/snapd
+	where := filepath.Join(rootDir, snapdMountPath)
+	if err := os.MkdirAll(where, 0755); err != nil {
+		return nil, err
+	}
+
+	removeMountpoint := func() {
```

Comment:
Why is retry-and-wait approach required? Can you add that as a code comment please.

Reply:
I've dropped it. As mentioned on the standup it was most likely a leftover from early version of this tool where some other problem (related to unmounting) prevented remove.